test/common/common/safe_memcpy_test.cc

```diff
@@ -0,0 +1,48 @@
+#include <array>
+
+#include "common/common/safe_memcpy.h"
+#include "common/grpc/common.h"
+#include "common/http/headers.h"
+#include "common/http/message_impl.h"
+#include "common/http/utility.h"
+
+#include "test/test_common/utility.h"
+
+#include "gtest/gtest.h"
+
+using testing::ElementsAre;
+using testing::Eq;
+
+namespace Envoy {
+
+TEST(SafeMemcpyTest, CopyUint8) {
+  const uint8_t src[] = {0, 1, 1, 2, 3, 5, 8, 13};
+  uint64_t dst;
+  safeMemcpy(&dst, &src);
+  Eq(dst == 282583128934413);
+}
+
+TEST(SafeMemcpyUnsafeSrcTest, CopyUint8Pointer) {
+  uint8_t* src = new uint8_t[8];
+  for (int i = 0; i < 8; ++i) {
+    src[i] = i;
+  }
+  uint8_t dst[8];
+  safeMemcpyUnsafeSrc(&dst, src);
+  ASSERT_THAT(dst, ElementsAre(0, 1, 2, 3, 4, 5, 6, 7));
```

Comment:
`EXPECT_THAT`

Usually we use the `ASSERT_*` forms if we need to exit the function if it fails, e.g. asserting the length of an array prior to checking an element value, so that the test binary doesn't crash when it fails. It's better otherwise to use EXPECT_* because then even if a test fails, subsequent checks may still run.

In this case, by using ASSERT, you would leak memory but skipping the delete if it fails. It doesn't matter that much really, but that's the common pattern.